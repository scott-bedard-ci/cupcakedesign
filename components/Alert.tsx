"use client"

import type React from "react"
import { X, AlertCircle, Info, CheckCircle, AlertTriangle } from "lucide-react"
import { cn } from "../lib/utils"
import { createComponent } from "../lib/createComponent"
import type { DismissableComponentProps } from "../lib/types"

// Alert variants
export type AlertVariant = "error" | "informational" | "success" | "warning"

// Alert properties
export interface AlertProps extends DismissableComponentProps, React.HTMLAttributes<HTMLDivElement> {
  /**
   * The title of the alert
   */
  title?: string
  /**
   * The message content of the alert
   */
  children?: React.ReactNode
  /**
   * The variant of the alert
   * @default "error"
   */
  variant?: AlertVariant
  /**
   * Whether to show the icon
   * @default true
   */
  showIcon?: boolean
  /**
   * Custom icon to display in the alert
   */
  icon?: React.ReactNode
  /**
   * Whether the alert should fill the width of its container
   * @default true
   */
  fill?: boolean
}

/**
 * Cupcake Alert component for displaying important messages to users
 */
export const Alert = createComponent<AlertProps>(
  {
    displayName: "Alert",
    defaultProps: {
      variant: "error",
      showIcon: true,
      showClose: true,
      fill: true,
    },
    dataAttributes: ["variant"],
  },
  (
    {
      title,
      children,
      variant = "error",
      showIcon = true,
      showClose = true,
      onDismiss,
      icon,
      className,
      fill = true,
      ...props
    },
    ref,
  ) => {
    // Map alert variant to styles
    const variantStyles = {
      error: {
        container: "border-alert-error bg-alert-error-background text-alert-error-foreground",
        icon: "text-alert-error",
        defaultIcon: <AlertCircle className="w-4 h-4 md:w-5 md:h-5" />,
      },
      informational: {
        container: "border-alert-info bg-alert-info-background text-alert-info-foreground",
        icon: "text-alert-info",
        defaultIcon: <Info className="w-4 h-4 md:w-5 md:h-5" />,
      },
      success: {
        container: "border-alert-success bg-alert-success-background text-alert-success-foreground",
        icon: "text-alert-success",
        defaultIcon: <CheckCircle className="w-4 h-4 md:w-5 md:h-5" />,
      },
      warning: {
        container: "border-alert-warning bg-alert-warning-background text-alert-warning-foreground",
        icon: "text-alert-warning",
        defaultIcon: <AlertTriangle className="w-4 h-4 md:w-5 md:h-5" />,
      },
    }

    // Get styles for current variant
    const currentStyle = variantStyles[variant]

    return (
      <div
        ref={ref}
        role="alert"
        className={cn(
          "flex items-start border rounded-md p-3 md:p-4",
          currentStyle.container,
          fill ? "w-full" : "inline-flex",
          className,
        )}
        {...props}
      >
        {showIcon && (
          <div className={cn("flex-shrink-0 mr-2 md:mr-3", currentStyle.icon)}>{icon || currentStyle.defaultIcon}</div>
        )}
        <div className="flex-1 min-w-0">
          {title && <h3 className="text-sm md:text-base font-semibold mb-0.5 md:mb-1">{title}</h3>}
          {children ? (
            <div className="text-xs md:text-sm">{children}</div>
          ) : (
            <div className="text-sm alert-content-empty"></div>
          )}
        </div>
        {showClose && (
          <button
            onClick={onDismiss}
            className="flex-shrink-0 ml-2 md:ml-3 hover:opacity-70 transition-opacity p-2 rounded-full flex items-center justify-center min-h-touch min-w-touch"
            aria-label="Dismiss alert"
          >
            <X className="w-4 h-4 md:w-5 md:h-5" />
          </button>
        )}
      </div>
    )
  },
)
