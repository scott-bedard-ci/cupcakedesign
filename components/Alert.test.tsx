import { render, screen, fireEvent } from "@testing-library/react"
import { Alert } from "./Alert"

describe("Alert Component", () => {
  // Rendering tests
  test("renders with default props", () => {
    render(<Alert>Test alert message</Alert>)
    expect(screen.getByRole("alert")).toBeInTheDocument()
    expect(screen.getByText("Test alert message")).toBeInTheDocument()
  })

  test("renders with title", () => {
    render(<Alert title="Alert Title">Test alert message</Alert>)
    expect(screen.getByText("Alert Title")).toBeInTheDocument()
    expect(screen.getByText("Test alert message")).toBeInTheDocument()
  })

  // Variant tests
  test.each(["error", "informational", "success", "warning"] as const)("renders %s variant correctly", (variant) => {
    render(
      <Alert variant={variant} title={`${variant.charAt(0).toUpperCase() + variant.slice(1)} Alert`}>
        This is a {variant} alert
      </Alert>,
    )

    const alertElement = screen.getByRole("alert")
    expect(alertElement).toHaveAttribute("data-cupcake-variant", variant)
    expect(screen.getByText(`${variant.charAt(0).toUpperCase() + variant.slice(1)} Alert`)).toBeInTheDocument()
    expect(screen.getByText(`This is a ${variant} alert`)).toBeInTheDocument()
  })

  // Icon tests
  test("renders with icon by default", () => {
    render(<Alert>Alert with icon</Alert>)
    // Check that there's an SVG element (icon)
    expect(document.querySelector("svg")).toBeInTheDocument()
  })

  test("does not render icon when showIcon is false", () => {
    render(<Alert showIcon={false}>Alert without icon</Alert>)
    // There should be no SVG element
    expect(document.querySelector("svg")).not.toBeInTheDocument()
  })

  // Close button tests
  test("calls onDismiss when close button is clicked", () => {
    const handleDismiss = jest.fn()
    render(<Alert onDismiss={handleDismiss}>Dismissible alert</Alert>)

    const dismissButton = screen.getByRole("button", { name: /dismiss alert/i })
    fireEvent.click(dismissButton)

    expect(handleDismiss).toHaveBeenCalledTimes(1)
  })

  test("does not render close button when showClose is false", () => {
    render(<Alert showClose={false}>Non-dismissible alert</Alert>)

    expect(screen.queryByRole("button", { name: /dismiss alert/i })).not.toBeInTheDocument()
  })

  // Fill width test
  test("has full width by default", () => {
    render(<Alert>Full width alert</Alert>)
    expect(screen.getByRole("alert")).toHaveClass("w-full")
  })

  test("has inline width when fill is false", () => {
    render(<Alert fill={false}>Inline width alert</Alert>)
    expect(screen.getByRole("alert")).toHaveClass("inline-flex")
    expect(screen.getByRole("alert")).not.toHaveClass("w-full")
  })

  // Content tests
  test("renders without title", () => {
    render(<Alert>Alert without title</Alert>)
    expect(screen.queryByRole("heading")).not.toBeInTheDocument()
    expect(screen.getByText("Alert without title")).toBeInTheDocument()
  })

  test("renders without children", () => {
    const { container } = render(<Alert title="Alert Title" />)
    expect(screen.getByText("Alert Title")).toBeInTheDocument()
    // The div that would contain children should be empty
    const textSmElement = container.querySelector(".alert-content-empty")
    expect(textSmElement).not.toBeNull()
    expect(textSmElement).toBeEmptyDOMElement()
  })

  // Custom icon test
  test("renders with custom icon", () => {
    const CustomIcon = () => <span data-testid="custom-icon">ðŸ””</span>
    render(<Alert icon={<CustomIcon />}>Alert with custom icon</Alert>)

    expect(screen.getByTestId("custom-icon")).toBeInTheDocument()
  })

  // Accessibility tests
  test("has appropriate ARIA attributes", () => {
    render(<Alert>Accessible alert</Alert>)

    const alertElement = screen.getByRole("alert")
    expect(alertElement).toBeInTheDocument()
  })

  // Snapshot test
  test("matches snapshot", () => {
    const { container } = render(
      <Alert title="Snapshot Test" variant="success" onDismiss={() => {}}>
        This is a snapshot test
      </Alert>,
    )

    // Update the snapshot
    expect(container).toMatchSnapshot()
  })
})
