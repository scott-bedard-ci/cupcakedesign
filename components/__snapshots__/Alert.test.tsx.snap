// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Alert Component matches snapshot 1`] = `
<div>
  <div
    class="flex items-start border rounded-md p-3 md:p-4 border-alert-success bg-alert-success-background text-alert-success-foreground w-full"
    data-cupcake-component="alert"
    data-cupcake-variant="success"
    role="alert"
  >
    <div
      class="flex-shrink-0 mr-2 md:mr-3 text-alert-success"
    >
      <svg
        class="w-4 h-4 md:w-5 md:h-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M22 11.08V12a10 10 0 1 1-5.93-9.14"
        />
        <polyline
          points="22 4 12 14.01 9 11.01"
        />
      </svg>
    </div>
    <div
      class="flex-1 min-w-0"
    >
      <h3
        class="text-sm md:text-base font-semibold mb-0.5 md:mb-1"
      >
        Snapshot Test
      </h3>
      <div
        class="text-xs md:text-sm"
      >
        This is a snapshot test
      </div>
    </div>
    <button
      aria-label="Dismiss alert"
      class="flex-shrink-0 ml-2 md:ml-3 hover:opacity-70 transition-opacity p-2 rounded-full flex items-center justify-center min-h-touch min-w-touch"
    >
      <svg
        class="w-4 h-4 md:w-5 md:h-5"
        fill="none"
        height="24"
        stroke="currentColor"
        stroke-linecap="round"
        stroke-linejoin="round"
        stroke-width="2"
        viewBox="0 0 24 24"
        width="24"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M18 6 6 18"
        />
        <path
          d="m6 6 12 12"
        />
      </svg>
    </button>
  </div>
</div>
`;
